### PR TITLE
🐛 fix(store): retain updatePolicy across container recreates (#496)

### LINKED ISSUE #496: I think I've found another bug related to the maturity of the updates.
### Drydock version

1.5.1

### What happened?

This has already happened to me in 1.5.1-rc.X, but it hasn't been resolved in 1.5.1.

I've had my doubts for a few days, but I've confirmed it, although I'm still not entirely sure how.

I have a DryDock instance on my system and a DryDock-Agent instance on a remote server.

I've added my agent to my system, and I can manage everything from there.

If I change the maturity level of a container in the remote agent and restart my instance, I lose the maturity level setting in that container.

I don't know if it's:
- Restarting the main DryDock instance.
- Restarting the remote agent.
- Because the image is being updated.
- Because "potatoes" 😂😂

This is causing updates as soon as a new image is released, without respecting the maturity level settings.

### What did you expect?

Maintain the config in remote containers.

### Steps to reproduce

I'm not entirely sure... I think it's when modifying the remote container's maturity and restarting the drydock instance.

### Relevant configuration

```yaml

```

### Logs

```text

```

### Installation method

Docker Compose

### Docker version

_No response_

### OS / Architecture

_No response_

### PR DESCRIPTION
**Pull request by scttbnsn**
Fixes #496.

## What's wrong

Container documents are keyed by Docker's **ephemeral container id**. A recreate — image pull, `docker compose up -d`, or one of our own action triggers firing — mints a brand-new id. Ingestion then misses the `getContainer(id)` lookup and falls to `insertContainer()` instead of `updateContainer()`.

`updateContainer()` merges the prior `updatePolicy` forward (`app/store/container.ts:826`). `insertContainer()` never has. The old document, which held the policy, is deleted with no carry-forward. So the replacement lands with `updatePolicy: undefined`, and `isUpdateSuppressed()` (`app/model/container.ts:557-561`) reads an absent policy as **no gating at all** rather than *default gating*.

The result is nasty: you get exactly one correctly-gated update, and the moment that update recreates the container your maturity gate evaporates permanently and silently. Every subsequent update fires with zero soak. The feature disables itself by working once.

That also explains the reporter's "only SOME containers reverted" — whether a container's id changed is a per-container condition. Restarted or untouched containers keep their id, hit `updateContainer`, and keep the policy.

## Provenance

**Pre-existing, not a 1.5.1 regression.** Per-container maturity policy landed in `b07195b3` (2026-03-14), first tagged **v1.4.1**. The `insertContainer`/`updateContainer` asymmetry itself dates to the initial commit `81fd6a41`. It is live in v1.4.x, v1.5.0, v1.5.1 GA, and `dev/v1.6`.

Ruled out along the way, so nobody re-litigates them: a `||`-vs-`??` falsy-coalesce bug (`resolveMaturityMinAgeDays` correctly uses `??`, and `parseMaturityMinAgeDays` rejects `0`), and any store migration (`app/store/migrate.ts` touches none of these fields).

## The fix

- **`app/store/container.ts`** — an `updatePolicyRetentionCache` keyed by `watcher::name`, stashed on `deleteContainer({replacementExpected: true})` and consumed by `insertContainer()`. It mirrors the existing `updateLifecycleCache` pattern, with two deliberate differences:
  - **Not gated behind `isLocalContainer`.** That exclusion (#386) exists because runtime *security state* is meaningless across agents. `updatePolicy` is durable controller-set config, and remote-agent containers are exactly the reported topology — they were strictly worse off before this.
  - **7-day TTL**, not the lifecycle cache's 30 minutes. A maturity soak encodes days-long user intent and a replacement container can lag its predecessor by a long image pull. A short TTL would only narrow the repro window, not close it.
  - An explicit incoming policy always wins, and the entry is consumed either way, so a stale policy can never attach to a later unrelated container.

- **`app/agent/AgentClient.ts`** — `pruneOldContainers()` deleted with **no replacement signal at all**, strictly worse than the local watcher's prune. It now mirrors that path: a stale-id entry whose name is still in the authoritative list is a replacement; anything else is a real removal.

  Worth calling out: flagging `replacementExpected` unconditionally here would have been a silent regression. `Hass.ts:477-497` reads that field off the container-removed event to decide whether to keep a state topic alive, so blanket-`true` would have orphaned Home Assistant discovery topics for genuinely deleted containers.

## Tests

Written first, and verified to fail against the unpatched source (8 of 11 store tests and the agent test fail without the fix).

- `app/store/container.test.ts` — carry-forward across a recreate; **the same test with `container.agent` set**, which is the #496 repro; no stash on a non-replacement delete; no stash without a policy; no stash for rollback containers; explicit payload wins; no cross-name leakage; TTL expiry; cap eviction; expired-first pruning; single consumption.
- `app/agent/AgentClient.test.ts` — flags a recreate, does **not** flag a genuine removal, and handles unnamed entries on both sides.

Full pre-push gate green: 100% coverage on `app` and `ui`, qlty clean, builds pass. 11,157 app tests pass.

## Release

Targeting a **`v1.5.2`** patch (RC first — the diff is small but it lands in `insertContainer`/`deleteContainer`, which every container in every deployment traverses on every scan). Forward-port to `dev/v1.6` follows the merge.

## Follow-up, not in scope here

`deriveContainerIdentityKey()` (`app/model/container.ts:937-958`) already computes a recreate-stable key and its own comment says it exists to "survive container recreates", but it is read nowhere in the store. Re-keying container documents on it is the real fix and is v1.7 debt. The retention cache is the surgical, patch-safe version.

## Changelog

- ✨ Added seven-day `updatePolicyRetentionCache` keyed by agent-scoped container identity (`deriveContainerIdentityKey()`), with TTL/max entries configurable via env.
- 🔧 Changed `insertContainer` to restore and then **consume** retained `updatePolicy` for recreated containers, applying the cached policy only when the incoming container does not explicitly include `updatePolicy`.
- 🔧 Changed `deleteContainer(..., { replacementExpected: true })` to stash the deleted container’s `updatePolicy` into the retention cache (skipping rollback container names / null policies).
- 🔧 Changed pruning (`pruneOldContainers`) to set `replacementExpected: true` when an agent-owned container is being recreated (same name still present), preserving discovery cleanup behavior for genuine removals.
- 🐛 Fixed loss of per-container maturity/updatePolicy settings for remote-agent and local containers across restarts/image updates/recreations.
- ✨ Added/expanded tests covering retention carry-forward, precedence, cross-agent isolation, expiration/eviction, single-use consumption, deletion/replacement scenarios, and identity-not-derivable edge cases.

- Verify cache TTL/max env overrides and parsing (`toPositiveInteger`) behave as intended.
- Confirm eviction/expiration ordering and single-use “consume then clear” semantics match expectations under concurrency.
- Validate identity-key derivation behavior when identities can’t be derived (cache retention vs consumption paths).
- Ensure rollback container name exclusion is correct and doesn’t block legitimate retention cases.

### CODE REVIEW

File: app/store/container.ts [hunk 1]
```diff
@@ -57,6 +61,24 @@ type UpdateLifecycleCacheEntry = {
...
+// isLocalContainer: the #386 exclusion exists because runtime state is meaningless across
+// agents, whereas updatePolicy is set on the controller and must survive for agent-owned
+// containers too — that is precisely the topology #496 was reported against.
+type UpdatePolicyRetentionCacheEntry = {
+  updatePolicy: unknown;
+  expiresAt: number;
+};
+
+const updatePolicyRetentionCache = new Map<string, UpdatePolicyRetentionCacheEntry>();
+
```
[the remark this answers is not in the document]
Comment (scttbnsn) [reply app/store/container.ts:92]: Good catch, this was a real bug and I've fixed it in efe542fe.

You're right about the collision. The reason it's specifically wrong here and not in the two neighbouring caches is that `securityStateCache` and `updateLifecycleCache` are both gated behind `isLocalContainer`, so they only ever hold local containers and `watcher::name` is sufficient. This cache deliberately drops that gate, because a remote agent's containers are exactly the case #496 was reported against, which is what makes the missing agent scope exploitable.

I went with `deriveContainerIdentityKey()` rather than hand-rolling the `${agent}::${watcher}::${name}` prefix. It already produces that key, it additionally prefers the compose `project/service` pair when those labels are present, and its own doc comment says it exists to survive container recreates. That's precisely this cache's requirement, and it means a compose-managed container that comes back under a different container name still inherits its policy.

Added three tests: cross-agent leak in both directions (agent A to agent B, and agent to local), and a compose-managed recreate under a changed name. Both leak tests fail against the old `watcher::name` key and pass against the new one, so I verified they actually catch what you found rather than just passing by construction.

**Review by biggest-littlest — APPROVED** [review]
LGTM. Root cause is well established and the agent-scoping catch is handled correctly.

**Review by ALARGECOMPANY — APPROVED** [review]
LGTM.
